File: core/src/com/jupiter/europa/entity/stats/characterclass/CharacterClass.java
/*
 * The MIT License
 *
 * Copyright 2014 Nathan Templon.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.jupiter.europa.entity.stats.characterclass;

import com.badlogic.ashley.core.Entity;
import com.badlogic.gdx.utils.Json;
import com.badlogic.gdx.utils.Json.Serializable;
import com.badlogic.gdx.utils.JsonValue;
import com.jupiter.europa.EuropaGame;
import com.jupiter.europa.entity.Families;
import com.jupiter.europa.entity.Mappers;
import com.jupiter.europa.entity.component.AttributesComponent;
import com.jupiter.europa.entity.messaging.RequestEffectAddMessage;
import com.jupiter.europa.entity.stats.AttributeSet;
import com.jupiter.europa.entity.stats.SkillSet.Skills;
import com.jupiter.europa.entity.trait.TraitPool;
import com.jupiter.europa.entity.trait.feat.FeatPool;
import com.jupiter.europa.entity.stats.race.Race;
import com.jupiter.europa.entity.trait.Trait;
import com.jupiter.europa.entity.trait.TraitPool.TraitPoolEvent;
import com.jupiter.europa.entity.trait.feat.Feat;
import com.jupiter.europa.util.Initializable;
import com.jupiter.ganymede.event.Event;
import com.jupiter.ganymede.event.Listener;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;
import org.reflections.Reflections;

/**
 *
 * @author Nathan Templon
 */
public abstract class CharacterClass implements Serializable, Initializable {

    // Constants
    public static final int MAX_LEVEL = 20;

    public static final Map<String, Class<? extends CharacterClass>> CLASS_LOOKUP = new HashMap<>();
    public static final String[] AVAILABLE_CLASSES = getAvailableClasses();

    private static final String LEVEL_KEY = "level";
    private static final String OWNER_ID_KEY = "owner-id";
    private static final String FEAT_POOL_KEY = "feats";
    public static final String AVAILABLE_SKILL_POINTS_KEY = "skill-points";


    // Static Methods
    public static CharacterClass getInstance(String className) {
        if (CLASS_LOOKUP.containsKey(className)) {
            Class<? extends CharacterClass> cc = CLASS_LOOKUP.get(className);
            try {
                return cc.newInstance();
            }
            catch (IllegalAccessException | InstantiationException ex) {
                return null;
            }

        }
        else {
            return null;
        }
    }

    public static int goodSave(int level) {
        return (int) (10.0 + (50.0 * ((level - 1.0) / (MAX_LEVEL - 1.0))));
    }

    public static int poorSave(int level) {
        return (int) (30.0 * ((level - 1.0) / (MAX_LEVEL - 1.0)));
    }

    public static int goodAttackBonus(int level) {
        double totalBonus = 0.0;
        double currentBonus = level;

        while (currentBonus > 0) {
            totalBonus += 5 * currentBonus;
            currentBonus -= 5;
        }

        return (int) Math.round(totalBonus);
    }

    public static int averageAttackBonus(int level) {
        double totalBonus = 0.0;
        double currentBonus = level * 0.75;

        while (currentBonus > 0) {
            totalBonus += 5 * currentBonus;
            currentBonus -= 5;
        }

        return (int) Math.round(totalBonus);
    }

    public static int poorAttackBonus(int level) {
        double totalBonus = 0.0;
        double currentBonus = level * 0.5;

        while (currentBonus > 0) {
            totalBonus += 5 * currentBonus;
            currentBonus -= 5;
        }

        return (int) Math.round(totalBonus);
    }

    private static String[] getAvailableClasses() {
        Reflections refl = new Reflections("com.jupiter.europa");
        Set<Class<? extends CharacterClass>> classes = refl.getSubTypesOf(CharacterClass.class);
        String[] classNames = new String[classes.size()];

        int i = 0;
        for (Class<? extends CharacterClass> charClass : classes) {
            String name = charClass.getSimpleName();
            classNames[i] = name;
            CLASS_LOOKUP.put(name, charClass);
            i++;
        }

        Arrays.sort(classNames);
        return classNames;
    }


    // Fields
    private final Event<LevelUpArgs> levelUp = new Event<>();

    private int level;
    private long ownerId = -1;
    private Entity owner;
    private final Set<TraitPool<? extends Trait>> abilityPools = new LinkedHashSet<>();
    private FeatPool featPool;
    private int availableSkillPoints;


    // Properties
    public final int getLevel() {
        return this.level;
    }

    public abstract int getHealthPerLevel();

    public abstract int getStartingHealth();

    public abstract int getSkillPointsPerLevel();

    public abstract int getAttackBonus();

    public abstract int getFortitude();

    public abstract int getReflexes();

    public abstract int getWill();

    public abstract Set<Skills> getClassSkills();

    public Set<TraitPool<? extends Trait>> getAbilityPools() {
        return this.abilityPools;
    }

    public final Entity getOwner() {
        return this.owner;
    }

    public final void setOwner(Entity owner) {
        this.owner = owner;

        this.abilityPools.stream().forEach((TraitPool<?> pool) -> {
            pool.setOwner(this.owner);
            pool.setAutoQualify(true);
            pool.reassesQualifications();
        });

//        if (Families.raced.matches(owner)) {
//            Race race = Mappers.race.get(owner).getRace();
//            if (race != null) {
//                if (this.featPool != null) {
//                    this.featPool.increaseCapacity(race.getNumberOfFeatsAtFirstLevel());
//                }
//            }
//        }
    }

    public FeatPool getFeatPool() {
        return this.featPool;
    }

    public int getMaxPointsPerSkill() {
        return 5 * (this.getLevel() + 3);
    }

    public int getAvailableSkillPoints() {
        return this.availableSkillPoints;
    }


    // Initialization
    public CharacterClass() {
        this.level = 1;

        // Feats
        this.featPool = new FeatPool();
        this.featPool.increaseCapacity(1);

        this.abilityPools.add(this.featPool);

        this.featPool.addSelectionListener(this::onFeatSelection);
    }

    @Override
    public void initialize() {
        if (this.ownerId > 0) {
            this.owner = EuropaGame.game.lastIdMapping.get(this.ownerId);
        }
        
        // Skills
        this.computeAvailableSkillPoints();

        if (this.owner != null) {
            this.abilityPools.stream().forEach((TraitPool<?> pool) -> {
                pool.setOwner(this.owner);
                pool.setAutoQualify(true);
                pool.reassesQualifications();
            });
        }
    }


    // Abstract Methods
    public abstract String getTextureSetName();


    // Public Methods
    public void levelUp() {
        if (this.level < MAX_LEVEL) {
            this.level++;

            if (this.level % 3 == 0) {
                this.featPool.increaseCapacity(1);
            }

            this.levelUp.dispatch(new LevelUpArgs(this, this.getLevel()));
        }
    }

    public boolean addLevelUpListener(Listener<LevelUpArgs> listener) {
        return this.levelUp.addListener(listener);
    }

    public boolean removeLevelUpListener(Listener<LevelUpArgs> listener) {
        return this.levelUp.removeListener(listener);
    }


    // Private Methods
    private void computeAvailableSkillPoints() {
        AttributesComponent comp = Mappers.attributes.get(this.getOwner());
        int intelligence = comp.getBaseAttributes().getAttribute(AttributeSet.Attributes.INTELLIGENCE);
        Race race = Mappers.race.get(this.getOwner()).getRace();
        int pointsPerLevel = this.getSkillPointsPerLevel() + race.getBonusSkillPoints() + intelligence / 2;
        this.availableSkillPoints = pointsPerLevel * (3 + this.getLevel());
    }

    private void onFeatSelection(TraitPoolEvent<Feat> event) {
        EuropaGame.game.getMessageSystem().publish(new RequestEffectAddMessage(this.owner, event.trait.getEffect()));
    }


    // Serializable (Json) Implementation
    @Override
    public void write(Json json) {
        json.writeValue(LEVEL_KEY, this.level);
        json.writeValue(OWNER_ID_KEY, this.owner.getId());
        json.writeValue(FEAT_POOL_KEY, this.featPool, this.featPool.getClass());
        json.writeValue(AVAILABLE_SKILL_POINTS_KEY, this.getAvailableSkillPoints());
    }

    @Override
    public void read(Json json, JsonValue jsonData) {
        if (jsonData.has(LEVEL_KEY)) {
            this.level = jsonData.getInt(LEVEL_KEY);
        }

        if (jsonData.has(OWNER_ID_KEY)) {
            this.ownerId = jsonData.getLong(OWNER_ID_KEY);
        }

        if (jsonData.has(FEAT_POOL_KEY)) {
            if (this.featPool != null) {
                this.abilityPools.remove(this.featPool);
            }
            this.featPool = json.fromJson(FeatPool.class, jsonData.get(FEAT_POOL_KEY).prettyPrint(EuropaGame.PRINT_SETTINGS));
            this.featPool.addSelectionListener(this::onFeatSelection);
            this.abilityPools.add(this.featPool);
        }

        if (jsonData.has(AVAILABLE_SKILL_POINTS_KEY)) {
            this.availableSkillPoints = jsonData.getInt(AVAILABLE_SKILL_POINTS_KEY);
        }
    }


    // Nested Classes
    public static class LevelUpArgs {

        // Fields
        public final int level;
        public final CharacterClass characterClass;


        // Initialization
        public LevelUpArgs(CharacterClass charClass, int level) {
            this.level = level;
            this.characterClass = charClass;
        }

    }

}

File: core/src/com/jupiter/europa/entity/stats/characterclass/CharacterClass.kt
/*
 * The MIT License
 *
 * Copyright 2015 Nathan Templon.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package com.jupiter.europa.entity.stats.characterclass

import com.badlogic.ashley.core.Entity
import com.badlogic.gdx.utils.Json
import com.badlogic.gdx.utils.Json.Serializable
import com.badlogic.gdx.utils.JsonValue
import com.jupiter.europa.EuropaGame
import com.jupiter.europa.entity.Mappers
import com.jupiter.europa.entity.effects.Effect
import com.jupiter.europa.entity.messaging.RequestEffectAddMessage
import com.jupiter.europa.entity.stats.AttributeSet
import com.jupiter.europa.entity.stats.SkillSet.Skills
import com.jupiter.europa.entity.traits.EffectPool
import com.jupiter.europa.entity.traits.EffectPool.TraitPoolEvent
import com.jupiter.europa.entity.traits.feat.Feat
import com.jupiter.europa.entity.traits.feat.FeatPool
import com.jupiter.europa.util.Initializable
import com.jupiter.ganymede.event.Event
import com.jupiter.ganymede.event.Listener
import org.reflections.Reflections
import java.util.LinkedHashSet

/**

 * @author Nathan Templon
 */
public abstract class CharacterClass : Serializable, Initializable {


    // Properties
    private val levelUp = Event<LevelUpArgs>()

    private var ownerId: Long = -1

    protected val abilityPoolsInternal: MutableSet<EffectPool<out Effect>> = LinkedHashSet<EffectPool<out Effect>>()

    public var level: Int = 0
        private set

    public var owner: Entity? = null
        get() = this.$owner
        set(value: Entity?) {
            this.$owner = value

            this.abilityPoolsInternal.forEach { pool ->
                pool.owner = this.owner
                pool.autoQualify = true
                pool.reassesQualifications()
            }
        }

    public val abilityPools: Set<EffectPool<out Effect>>
        get() = this.abilityPoolsInternal

    public val maxPointsPerSkill: Int
        get() = 5 * (this.level + 3)

    public var featPool: FeatPool = FeatPool()
        private set

    public var availableSkillPoints: Int = 0
        private set

    public abstract val healthPerLevel: Int
    public abstract val startingHealth: Int
    public abstract val skillPointsPerLevel: Int
    public abstract val attackBonus: Int
    public abstract val fortitude: Int
    public abstract val reflexes: Int
    public abstract val will: Int
    public abstract val stamina: Int
    public abstract val mana: Int
    public abstract val manaAttribute: AttributeSet.Attributes
    public abstract val aether: Int
    public abstract val aetherAttribute: AttributeSet.Attributes
    public abstract val classSkills: Set<Skills>
    public abstract val textureSetName: String


    init {
        this.level = 1

        // Feats
        this.featPool.increaseCapacity(1)
        this.abilityPoolsInternal.add(this.featPool)
        this.featPool.addSelectionListener(Listener { args -> this.onFeatSelection(args) })
    }

    override fun initialize() {
        if (this.ownerId > 0) {
            this.owner = EuropaGame.game.lastIdMapping.get(this.ownerId)
        }

        // Skills
        this.computeAvailableSkillPoints()

        if (this.owner != null) {
            this.abilityPoolsInternal.forEach { pool ->
                pool.owner = this.owner
                pool.autoQualify = true
                pool.reassesQualifications()
            }
        }
    }

    public open fun onFirstCreation() {
    }

    // Public Methods
    public open fun levelUp() {
        if (this.level < MAX_LEVEL) {
            this.level++

            if (this.level % 3 == 0) {
                this.featPool!!.increaseCapacity(1)
            }

            this.levelUp.dispatch(LevelUpArgs(this, this.level))
        }
    }

    public fun addLevelUpListener(listener: Listener<LevelUpArgs>): Boolean {
        return this.levelUp.addListener(listener)
    }

    public fun removeLevelUpListener(listener: Listener<LevelUpArgs>): Boolean {
        return this.levelUp.removeListener(listener)
    }


    // Private Methods
    private fun computeAvailableSkillPoints() {
        val comp = Mappers.attributes[this.owner]
        val intelligence = comp.baseAttributes.getAttribute(AttributeSet.Attributes.INTELLIGENCE)
        val race = Mappers.race.get(this.owner).race
        val pointsPerLevel = this.skillPointsPerLevel +
                (race?.bonusSkillPoints ?: 0) +
                intelligence / 2
        this.availableSkillPoints = pointsPerLevel * (3 + this.level)
    }

    private fun onFeatSelection(event: TraitPoolEvent<Feat>) {
        if (this.owner != null) {
            EuropaGame.game.messageSystem.publish(RequestEffectAddMessage(this.owner!!, event.`trait`))
        }
    }


    // Serializable (Json) Implementation
    override fun write(json: Json) {
        json.writeValue(LEVEL_KEY, this.level)
        json.writeValue(OWNER_ID_KEY, this.owner!!.getId())
        json.writeValue(FEAT_POOL_KEY, this.featPool, this.featPool!!.javaClass)
        json.writeValue(AVAILABLE_SKILL_POINTS_KEY, this.availableSkillPoints)
    }

    override fun read(json: Json, jsonData: JsonValue) {
        if (jsonData.has(LEVEL_KEY)) {
            this.level = jsonData.getInt(LEVEL_KEY)
        }

        if (jsonData.has(OWNER_ID_KEY)) {
            this.ownerId = jsonData.getLong(OWNER_ID_KEY)
        }

        if (jsonData.has(FEAT_POOL_KEY)) {
            this.abilityPoolsInternal.remove(this.featPool)
            this.featPool = json.fromJson(javaClass<FeatPool>(), jsonData.get(FEAT_POOL_KEY).prettyPrint(EuropaGame.PRINT_SETTINGS))
            this.featPool.addSelectionListener(Listener { args -> this.onFeatSelection(args) })
            this.abilityPoolsInternal.add(this.featPool)
        }

        if (jsonData.has(AVAILABLE_SKILL_POINTS_KEY)) {
            this.availableSkillPoints = jsonData.getInt(AVAILABLE_SKILL_POINTS_KEY)
        }
    }


    // Nested Classes
    public class LevelUpArgs(public val characterClass: CharacterClass, public val level: Int)

    companion object {

        // Constants
        public val MAX_LEVEL: Int = 20

        public val CLASS_LOOKUP: Map<String, Class<out CharacterClass>> = getClassMapping()
        public val AVAILABLE_CLASSES: List<String> = CLASS_LOOKUP.keySet().toList().sort()

        private val LEVEL_KEY = "level"
        private val OWNER_ID_KEY = "owner-id"
        private val FEAT_POOL_KEY = "feats"
        public val AVAILABLE_SKILL_POINTS_KEY: String = "skill-points"


        // Static Methods
        public fun getInstance(className: String): CharacterClass? {
            if (CLASS_LOOKUP.containsKey(className)) {
                val cc = CLASS_LOOKUP.get(className)
                try {
                    return cc?.newInstance()
                } catch (ex: IllegalAccessException) {
                    return null
                } catch (ex: InstantiationException) {
                    return null
                }


            } else {
                return null
            }
        }

        public fun goodSave(level: Int): Int {
            return (10.0 + (50.0 * ((level - 1.0) / (MAX_LEVEL - 1.0)))).toInt()
        }

        public fun poorSave(level: Int): Int {
            return (30.0 * ((level - 1.0) / (MAX_LEVEL - 1.0))).toInt()
        }

        public fun goodAttackBonus(level: Int): Int {
            var totalBonus = 0.0
            var currentBonus = level.toDouble()

            while (currentBonus > 0) {
                totalBonus += 5 * currentBonus
                currentBonus -= 5.0
            }

            return Math.round(totalBonus).toInt()
        }

        public fun averageAttackBonus(level: Int): Int {
            var totalBonus = 0.0
            var currentBonus = level * 0.75

            while (currentBonus > 0) {
                totalBonus += 5 * currentBonus
                currentBonus -= 5.0
            }

            return Math.round(totalBonus).toInt()
        }

        public fun poorAttackBonus(level: Int): Int {
            var totalBonus = 0.0
            var currentBonus = level * 0.5

            while (currentBonus > 0) {
                totalBonus += 5 * currentBonus
                currentBonus -= 5.0
            }

            return Math.round(totalBonus).toInt()
        }

        private fun getClassMapping(): Map<String, Class<out CharacterClass>> {
            val refl = Reflections("com.jupiter.europa")
            val classes = refl.getSubTypesOf(javaClass<CharacterClass>())

            return classes.map { type ->
                Pair(type.getSimpleName(), type)
            }.toMap()
        }
    }

}
